File: ParkingGameCpp/src/car.cpp
#include "car.h"

#include <godot_cpp/classes/box_mesh.hpp>
#include <godot_cpp/classes/box_shape3d.hpp>
#include <godot_cpp/classes/collision_shape3d.hpp>
#include <godot_cpp/classes/cylinder_mesh.hpp>
#include <godot_cpp/classes/mesh_instance3d.hpp>
#include <godot_cpp/core/math.hpp>

namespace godot {

float CppCar::get_forward_speed() const {
    return -get_global_transform().basis.get_column(2).dot(get_linear_velocity());
}

std::vector<Vector3> CppCar::corners() const {
    const Basis &b = get_global_transform().basis;
    const Vector3 o = get_global_position();
    Vector3 hw = b.get_column(0) * (SEDAN_WID * 0.5f);
    Vector3 hf = b.get_column(2) * (SEDAN_LEN * 0.5f);
    return { o + hw + hf, o + hw - hf, o - hw - hf, o - hw + hf };
}

void CppCar::set_brake_lights(bool on) {
    if (_tail.is_valid()) {
        _tail->set_emission_energy_multiplier(on ? 4.5f : 1.6f);
    }
}

VehicleWheel3D *CppCar::make_wheel(const Vector3 &pos, bool steering, bool traction) {
    VehicleWheel3D *w = memnew(VehicleWheel3D);
    w->set_position(pos);
    w->set_radius(0.34f);
    w->set_suspension_rest_length(0.35f);
    w->set_suspension_travel(0.20f);
    w->set_suspension_stiffness(55.0f);
    w->set_damping_compression(6.0f);
    w->set_damping_relaxation(7.5f);
    w->set_suspension_max_force(60000.0f);
    w->set_friction_slip(11.0f);
    w->set_roll_influence(0.08f);
    w->set_use_as_steering(steering);
    w->set_use_as_traction(traction);
    add_child(w);
    return w;
}

void CppCar::_ready() {
    set_mass(1200.0f);
    set_can_sleep(false);

    CollisionShape3D *shape = memnew(CollisionShape3D);
    Ref<BoxShape3D> box;
    box.instantiate();
    box->set_size(Vector3(SEDAN_WID, 1.05f, SEDAN_LEN));
    shape->set_shape(box);
    shape->set_position(Vector3(0, 0.55f, 0));
    add_child(shape);

    _fl = make_wheel(Vector3(-0.80f, 0.10f, -1.48f), true, false);
    _fr = make_wheel(Vector3(0.80f, 0.10f, -1.48f), true, false);
    _rl = make_wheel(Vector3(-0.80f, 0.10f, 1.48f), false, true);
    _rr = make_wheel(Vector3(0.80f, 0.10f, 1.48f), false, true);

    build_visuals();
}

void CppCar::build_visuals() {
    Ref<StandardMaterial3D> red;
    red.instantiate();
    red->set_albedo(Color(0.82f, 0.27f, 0.24f));
    red->set_roughness(0.32f);
    red->set_metallic(0.12f);

    Ref<StandardMaterial3D> dark;
    dark.instantiate();
    dark->set_albedo(Color(0.14f, 0.15f, 0.17f));
    dark->set_roughness(0.5f);

    Ref<StandardMaterial3D> glass;
    glass.instantiate();
    glass->set_albedo(Color(0.10f, 0.13f, 0.16f, 0.85f));
    glass->set_transparency(BaseMaterial3D::TRANSPARENCY_ALPHA);
    glass->set_roughness(0.08f);
    glass->set_metallic(0.9f);

    Ref<StandardMaterial3D> tire;
    tire.instantiate();
    tire->set_albedo(Color(0.09f, 0.09f, 0.10f));
    tire->set_roughness(0.95f);

    Ref<StandardMaterial3D> light;
    light.instantiate();
    light->set_albedo(Color(1.0f, 0.97f, 0.85f));
    light->set_feature(BaseMaterial3D::FEATURE_EMISSION, true);
    light->set_emission(Color(1.0f, 0.95f, 0.8f));
    light->set_emission_energy_multiplier(1.6f);

    _tail.instantiate();
    _tail->set_albedo(Color(0.9f, 0.1f, 0.1f));
    _tail->set_feature(BaseMaterial3D::FEATURE_EMISSION, true);
    _tail->set_emission(Color(1.0f, 0.12f, 0.1f));
    _tail->set_emission_energy_multiplier(1.6f);

    auto mesh = [](const Vector3 &size, const Ref<StandardMaterial3D> &mat,
                   const Vector3 &pos) {
        MeshInstance3D *m = memnew(MeshInstance3D);
        Ref<BoxMesh> bm;
        bm.instantiate();
        bm->set_size(size);
        m->set_mesh(bm);
        m->set_material_override(mat);
        m->set_position(pos);
        return m;
    };

    add_child(mesh(Vector3(SEDAN_WID, 0.55f, SEDAN_LEN), red, Vector3(0, 0.52f, 0)));
    add_child(mesh(Vector3(SEDAN_WID - 0.14f, 0.30f, 2.16f), glass, Vector3(0, 0.95f, 0.25f)));
    add_child(mesh(Vector3(SEDAN_WID - 0.2f, 0.20f, 2.0f), red, Vector3(0, 1.20f, 0.22f)));

    for (float z : { -SEDAN_LEN / 2.0f - 0.03f, SEDAN_LEN / 2.0f + 0.03f }) {
        add_child(mesh(Vector3(SEDAN_WID + 0.06f, 0.30f, 0.12f), dark, Vector3(0, 0.38f, z)));
    }
    for (float x : { -0.55f, 0.55f }) {
        add_child(mesh(Vector3(0.30f, 0.12f, 0.06f), light, Vector3(x, 0.55f, -SEDAN_LEN / 2.0f - 0.02f)));
        add_child(mesh(Vector3(0.34f, 0.18f, 0.10f), _tail, Vector3(x, 0.56f, SEDAN_LEN / 2.0f + 0.04f)));
    }

    for (float x : { -0.80f, 0.80f }) {
        for (float z : { -1.48f, 1.48f }) {
            Node3D *pivot = memnew(Node3D);
            pivot->set_position(Vector3(x, 0.10f, z));
            MeshInstance3D *wheel = memnew(MeshInstance3D);
            Ref<CylinderMesh> cm;
            cm.instantiate();
            cm->set_top_radius(0.34f);
            cm->set_bottom_radius(0.34f);
            cm->set_height(0.26f);
            wheel->set_mesh(cm);
            wheel->set_material_override(tire);
            wheel->set_rotation(Vector3(0, 0, Math::PI / 2.0f)); // cylinder axis -> X
            pivot->add_child(wheel);
            add_child(pivot);
            if (z < 0) {
                if (x < 0) _fl_pivot = pivot;
                else _fr_pivot = pivot;
            }
        }
    }
}

void CppCar::_physics_process(double delta) {
    float dt = static_cast<float>(delta);
    float speed = get_forward_speed();

    // engine by gear (sign note: body-level force positive pushes +Z / tail)
    float engine = 0.0f;
    switch (current_gear) {
        case GEAR_D:
            if (throttle > 0.01f) {
                if (speed < 13.0f) {
                    float fade = Math::clamp(1.0f - speed / 13.0f, 0.0f, 1.0f);
                    engine = -3400.0f * throttle * (0.4f + 0.6f * fade);
                }
            } else if (speed > 0.3f) {
                engine = 550.0f; // coasting forward -> drag toward +Z
            }
            break;
        case GEAR_R:
            if (throttle > 0.01f) {
                if (speed > -6.5f) {
                    float fade = Math::clamp(1.0f + speed / 6.5f, 0.0f, 1.0f);
                    engine = 2400.0f * throttle * (0.4f + 0.6f * fade);
                }
            } else if (speed < -0.3f) {
                engine = -550.0f;
            }
            break;
        default:
            break;
    }
    set_engine_force(engine);

    float brake = brake_input * 7.0f;
    if (handbrake) {
        brake = Math::max(brake, 7.0f * 0.9f);
    }
    if (brake_input > 0.5f && Math::abs(speed) < 0.1f && throttle < 0.01f) {
        set_linear_damp(8.0f);
        brake = 7.0f;
    } else {
        set_linear_damp(0.0f);
    }
    set_brake(brake);

    // speed-sensitive steering + rate-limited actuator
    float max_steer = Math::deg_to_rad(34.0f) *
        Math::clamp(1.0f - Math::abs(speed) / 16.0f, 0.3f, 1.0f);
    float target = Math::clamp(steer_input, -1.0f, 1.0f) * max_steer;
    _steer = Math::lerp(_steer, target, Math::clamp(7.0f * dt, 0.0f, 1.0f));
    set_steering(_steer);
    if (_fl_pivot) _fl_pivot->set_rotation(Vector3(0, _steer, 0));
    if (_fr_pivot) _fr_pivot->set_rotation(Vector3(0, _steer, 0));
}

} // namespace godot

File: ParkingGame/scripts/Car.cs
using Godot;

namespace ParkingGame;

/// <summary>
/// The player car — a VehicleBody3D with a manual gearbox (R/N/D) and speed-sensitive
/// steering. Controls are written into the public fields by Game every physics tick
/// (keyboard or the --demo script); the physics integration is all Godot's raycast
/// vehicle: per-wheel suspension, friction slip, traction and brake forces.
/// </summary>
public partial class Car : VehicleBody3D
{
    public enum Gear { N, D, R }
    public Gear CurrentGear = Gear.N;

    // ---- control inputs (0..1 / -1..1), written externally each tick ----
    public float Throttle;    // 0..1
    public float BrakeInput;  // 0..1
    public float SteerInput;  // -1..1, +1 = wheels left (Godot steering convention)
    public bool Handbrake;

    public const float BodyLen = 4.6f;
    public const float BodyWid = 1.8f;

    // ---- tuning ----
    public float MaxEngineForce = 3400f;   // N total at driven wheels (forward)
    public float MaxReverseForce = 2400f;
    public float EngineBrakeForce = 550f;  // coasting drag from the drivetrain
    public float MaxBrake = 7.0f;           // wheel brake strength (tuned empirically)
    public float MaxSteerDeg = 34f;
    public float SteerFadeSpeed = 16f;     // steering clamps down as speed climbs
    public float SteerLerpRate = 7f;
    public float TopSpeedD = 13f;          // m/s (~47 km/h — a parking lot)
    public float TopSpeedR = 6.5f;

    public float ForwardSpeed => -GlobalTransform.Basis.Z.Dot(LinearVelocity);

    private VehicleWheel3D _fl, _fr, _rl, _rr;
    private Node3D _flPivot, _frPivot;     // visual front wheels (steer with input)
    private float _steer;
    private float _collideCooldown;
    private readonly System.Collections.Generic.HashSet<Node> _recentHits = new();

    public int CollisionCount { get; private set; }
    public bool CollisionThisTick { get; private set; }

    public override void _Ready()
    {
        Mass = 1200f;
        CanSleep = false;

        // chassis collision
        var shape = new CollisionShape3D
        {
            Shape = new BoxShape3D { Size = new Vector3(BodyWid, 1.05f, BodyLen) },
            Position = new Vector3(0, 0.55f, 0),
        };
        AddChild(shape);

        // ---- wheels: front steer, rear drive (classic RWD parking feel) ----
        _fl = MakeWheel(new Vector3(-0.80f, 0.10f, -1.48f), steering: true);
        _fr = MakeWheel(new Vector3(0.80f, 0.10f, -1.48f), steering: true);
        _rl = MakeWheel(new Vector3(-0.80f, 0.10f, 1.48f), steering: false, traction: true);
        _rr = MakeWheel(new Vector3(0.80f, 0.10f, 1.48f), steering: false, traction: true);

        BuildVisuals();

        BodyEntered += OnBodyEntered;
    }

    private VehicleWheel3D MakeWheel(Vector3 pos, bool steering, bool traction = false)
    {
        var w = new VehicleWheel3D
        {
            Position = pos,
            WheelRadius = 0.34f,
            WheelRestLength = 0.35f,
            SuspensionTravel = 0.20f,
            SuspensionStiffness = 55f,
            DampingCompression = 6.0f,
            DampingRelaxation = 7.5f,
            SuspensionMaxForce = 60000f,
            WheelFrictionSlip = 11f,
            WheelRollInfluence = 0.08f,
            UseAsSteering = steering,
            UseAsTraction = traction,
        };
        AddChild(w);
        return w;
    }

    private void BuildVisuals()
    {
        var red = new StandardMaterial3D { AlbedoColor = new Color(0.82f, 0.27f, 0.24f) };
        var dark = new StandardMaterial3D { AlbedoColor = new Color(0.14f, 0.15f, 0.17f) };
        var tire = new StandardMaterial3D { AlbedoColor = new Color(0.09f, 0.09f, 0.10f) };

        var body = new MeshInstance3D
        {
            Mesh = new BoxMesh { Size = new Vector3(BodyWid, 0.55f, BodyLen) },
            MaterialOverlay = red,
            Position = new Vector3(0, 0.52f, 0),
        };
        AddChild(body);
        var cabin = new MeshInstance3D
        {
            Mesh = new BoxMesh { Size = new Vector3(BodyWid - 0.2f, 0.48f, 2.1f) },
            MaterialOverride = dark,
            Position = new Vector3(0, 1.02f, 0.25f),
        };
        AddChild(cabin);

        foreach (var (x, z) in new[] { (-0.80f, -1.48f), (0.80f, -1.48f), (-0.80f, 1.48f), (0.80f, 1.48f) })
        {
            var pivot = new Node3D { Position = new Vector3(x, 0.10f, z) };
            var mesh = new MeshInstance3D
            {
                Mesh = new CylinderMesh { TopRadius = 0.34f, BottomRadius = 0.34f, Height = 0.26f },
                MaterialOverride = tire,
                Rotation = new Vector3(0, 0, Mathf.Pi / 2f), // cylinder axis → X
            };
            pivot.AddChild(mesh);
            AddChild(pivot);
            if (z < 0)
            {
                if (x < 0) _flPivot = pivot; else _frPivot = pivot;
            }
        }
    }

    public void SelectGear(Gear g) => CurrentGear = g;

    public void ResetTo(Vector3 pos, float yawDeg)
    {
        CurrentGear = Gear.N;
        Throttle = BrakeInput = SteerInput = 0;
        Handbrake = false;
        _steer = 0;
        LinearVelocity = Vector3.Zero;
        AngularVelocity = Vector3.Zero;
        Steering = 0;
        EngineForce = 0;
        Brake = 1;
        GlobalPosition = pos;
        Rotation = new Vector3(0, Mathf.DegToRad(yawDeg), 0);
        CollisionCount = 0;
        _recentHits.Clear();
        Sleeping = false;
    }

    public override void _PhysicsProcess(double delta)
    {
        float dt = (float)delta;
        float speed = ForwardSpeed;

        // ---- engine by gear (realistic curve: strong at low speed, capped) ----
        // Sign note (verified empirically on 4.7.1): body-level EngineForce is
        // POSITIVE-toward-+Z — the opposite of the Bullet raycast-vehicle docs.
        // Forward (-Z) drive therefore needs NEGATIVE force values here.
        float engine = 0f;
        switch (CurrentGear)
        {
            case Gear.D:
                if (Throttle > 0.01f)
                {
                    if (speed < TopSpeedD)
                    {
                        float fade = Mathf.Clamp(1f - speed / TopSpeedD, 0f, 1f);
                        engine = -MaxEngineForce * Throttle * (0.4f + 0.6f * fade);
                    }
                }
                else if (speed > 0.3f)
                {
                    engine = EngineBrakeForce; // coasting forward → drag toward +Z
                }
                break;
            case Gear.R:
                if (Throttle > 0.01f)
                {
                    if (speed > -TopSpeedR)
                    {
                        float fade = Mathf.Clamp(1f + speed / TopSpeedR, 0f, 1f);
                        engine = MaxReverseForce * Throttle * (0.4f + 0.6f * fade);
                    }
                }
                else if (speed < -0.3f)
                {
                    engine = -EngineBrakeForce; // coasting backward → drag toward -Z
                }
                break;
            case Gear.N:
                break; // freewheel — no drive, no engine brake
        }
        EngineForce = engine;

        // ---- brakes (all wheels; handbrake clamps to a strong stop) ----
        float brake = BrakeInput * MaxBrake;
        if (Handbrake)
            brake = Mathf.Max(brake, MaxBrake * 0.9f);
        // holding the brake at a standstill pins the car (no creep)
        if (BrakeInput > 0.5f && Mathf.Abs(speed) < 0.1f && Throttle < 0.01f)
        {
            LinearDamp = 8f;
            brake = MaxBrake;
        }
        else
        {
            LinearDamp = 0f;
        }
        Brake = brake;

        // ---- steering: speed-sensitive limit + rate-limited actuator ----
        float maxSteer = Mathf.DegToRad(MaxSteerDeg) *
                         Mathf.Clamp(1f - Mathf.Abs(speed) / SteerFadeSpeed, 0.3f, 1f);
        float target = Mathf.Clamp(SteerInput, -1f, 1f) * maxSteer;
        _steer = Mathf.Lerp(_steer, target, Mathf.Clamp(SteerLerpRate * dt, 0f, 1f));
        Steering = _steer;

        if (_flPivot != null) _flPivot.Rotation = new Vector3(0, _steer, 0);
        if (_frPivot != null) _frPivot.Rotation = new Vector3(0, _steer, 0);

        // ---- collision bookkeeping (walls / parked cars / cones — not the floor) ----
        CollisionThisTick = false;
        _collideCooldown -= dt;
        if (_collideCooldown <= 0f)
        {
            _recentHits.Clear();
        }
    }

    private void OnBodyEntered(Node body)
    {
        if (body.Name == "Floor" || body == this)
            return;
        // a fresh burst within the cooldown window only counts once
        if (_recentHits.Add(body))
        {
            CollisionCount++;
            CollisionThisTick = true;
            _collideCooldown = 0.5f;
        }
    }

    /// <summary>Four chassis corners in world space, used by the slot check.</summary>
    public Vector3[] Corners()
    {
        var b = GlobalTransform.Basis;
        var o = GlobalPosition;
        Vector3 hw = b.X * (BodyWid / 2f);
        Vector3 hf = b.Z * (BodyLen / 2f);
        return new[]
        {
            o + hw + hf, o + hw - hf, o - hw + hf, o - hw - hf,
        };
    }
}
